### ISSUE TITLE
[Issue]: wrong Datasource 'get' method behaviour after appended/prepended elements 

### ISSUE THREAD

**Issue by IgorVNC** [dhilt/ngx-ui-scroll#219]
Hello @dhilt 

I found one issue with the lib, related to wrong Datasource 'get' method behaviour after appended/prepended elements 

I have prepared a Stackblitz demo so it's easy to reproduce it 
https://stackblitz.com/edit/angular-ngx-ui-scroll-crash-on-append-x5e1pr?file=src/app/app.component.ts

Steps to reproduce:
1) Use inverted & reversed datasource flow (I believe the issue is also reproducible with normal direction)
2) Once ran the demo, the Datasource 'get' method requested the following range:
-9 -> 0
1 -> 10
11 -> 20
21 -> 30
3) append 25 items
4) start scrolling to top 

Result: 

Instead of requesting next batch (30+25) -> (40+25), it requests 31 -> 40 which is wrong, 
so same elements will be displayed twice (see screenshot)

<img width="325" alt="Screenshot 2020-10-13 at 17 40 48" src="https://user-images.githubusercontent.com/57942649/95875968-43872180-0d7b-11eb-84a4-a072bd34cdcb.png">

It seems the Datasource 'get' method is not aware of newly appended/prepended elements, and does not adjust its internals counters, so the issue pops up  

Plese let me know if any other details are required to provide 

Thank you in advance

**Comment by dhilt** [comment]
@IgorVNC Thanks for the demo, it makes the problem clear. And this is not a bug, this is about the difficulties of the Datasource API applied to the case of inversion. Point 4 reveals the details and the solution.

1. Update please ngx-ui-scroll to v1.8.5 as it has some stability improvements related to `Adapter.relax`.

2. `Adapter.relax` + `Adapter.isLoading`, this double protection is not necessary. The `isLoading` flag will always be `false` inside the `relax` callback. If it's not true in some case and you can reproduce it, I would really appreciate if you could open a separate issue for this.

3. If you are not at the bottom line of the viewport and there are some virtual items after the last rendered one, then appending should be done virtually. You may read about it [here](https://dhilt.github.io/ngx-ui-scroll/#/adapter#append-prepend-sync). This behavior is achieved via `eof` setting:

```js
  this.datasource.adapter.append({ items, eof: true });
```

4. The main problem is the indexes inconsistency. Initially you have 100 items with natural indexes from 0 to 99. But the datasource indexes are different. Due to inversion, you have a list of items with indexes from -100 to -1. The mapping you implemented in the `Datasource.get` method works fine until you break it by appending new zero-positive items. After appending, say, 25 items, your Datasource consists of 125 items with indexes from -100 to 24. But natural indexes of the data array are strongly positive: from 0 to 124. You have to manage this shift at the `Datasource.get` level in order to provide correct mapping after append:

```js
  const start = this.addedCount - index - count;
```

The updated demo is here: https://stackblitz.com/edit/angular-ngx-ui-scroll-crash-on-append-fork

**Comment by IgorVNC** [comment]
@dhilt Thnaks for your reply, please see my comments:

1 - will do 

2 - indeed this is a real issue and in big projects it can be easily faced w/o much effort. By big projects I mean some complex UI components to render, which requires more CPU/render time to display. For example, I have a chatting app, similar to Slack, with complex UI components for chat messages, with dozens sub elements. And sometimes we face such an issue with 'relax'. I found a similar issue https://github.com/dhilt/ngx-ui-scroll/issues/187

Also, please find a stackblitz demo 
https://stackblitz.com/edit/angular-ngx-ui-scroll-relax-issue

I was not able to find a simple way to repro it using these simple examples, 
but here are the clear steps re how to reproduce a crash:
1) press 'Append' button 5 times, so it will start 5 threads to append elements
2) then just scroll to top and you will get the issue

<img width="1300" alt="Screenshot 2020-10-13 at 22 18 41" src="https://user-images.githubusercontent.com/57942649/95905885-100bbd80-0da2-11eb-9555-c9171284f603.png">

And then, if you uncomment that 'isLoading' check return, then I will start work well. 
So I'm looking forward to find a way re how to mitigate it. 

Basically, 2 ideas here:
1) fix 'relax' method
2) fix 'relax' method and create a new set of 'safe' methods, e.g. 'safeAppend', 'safePrepend', 'safeReload', where internally they will be using a fixed version of 'relax' method and do all the durty job, so the end users do not need to care about it all

3 - thanks. I know about this API, but was not able to properly undderstand its usage, will try here again

4 - I had same idea in my mind. The only drawback here I see is that now we need to handle the case of reseting this shift, e.g. when we do 'reload' or other operations which lead to reloading/reseting the datasource internals. It all must be done carefylly. 

Thnaks for your help!

**Comment by IgorVNC** [comment]
As for the info, `this.addedCount` shift should be added to  lastVisible/firstVisible values as well to get proper values 


```
let lastVisibleIndex = this.datasource.adapter.lastVisible.$index;
let firstVisibleIndex = this.datasource.adapter.firstVisible.$index;
```

**Comment by dhilt** [comment]
Great! I think we can continue the investigation in the issue https://github.com/dhilt/ngx-ui-scroll/issues/220. Closing this as the original (datasource) issue is resolved.
